dll/win32/imm32/ctf.c

```diff
@@ -17,27 +17,139 @@ WINE_DEFAULT_DEBUG_CHANNEL(imm);
+
+BOOL Imm32CheckAndApplyAppCompat(ULONG dwReason, LPCWSTR pszAppName)
+{
+    HINSTANCE hinstApphelp;
+    FN_ApphelpCheckIME pApphelpCheckIME;
+
+    if (BaseCheckAppcompatCache(pszAppName, INVALID_HANDLE_VALUE, NULL, &dwReason))
+        return TRUE;
+
+    hinstApphelp = GetModuleHandleW(L"apphelp.dll");
+    if (!hinstApphelp)
+    {
+        hinstApphelp = LoadLibraryW(L"apphelp.dll");
+        if (!hinstApphelp)
+            return TRUE;
+    }
+
+    pApphelpCheckIME = (FN_ApphelpCheckIME)GetProcAddress(hinstApphelp, "ApphelpCheckIME");
+    if (!pApphelpCheckIME)
+        return TRUE;
+
+    return pApphelpCheckIME(pszAppName);
+}
+
 HMODULE APIENTRY Imm32LoadCtfIme(VOID)
 {
-    return NULL;
+    BOOL bSuccess = FALSE;
+    IMEINFOEX ImeInfoEx;
+    WCHAR szImeFile[MAX_PATH];
+
+    RtlEnterCriticalSection(&gcsImeDpi);
+    do
+    {
+        if (g_hCtfIme)
+            break;
+
+        /*
+         * NOTE: (HKL)0x04090409 is English US keyboard (default).
+         * The Cicero keyboard logically uses English US keyboard.
+         */
+        if (!ImmLoadLayout((HKL)ULongToHandle(0x04090409), &ImeInfoEx))
+            break;
+
+        Imm32GetSystemLibraryPath(szImeFile, _countof(szImeFile), ImeInfoEx.wszImeFile);
+
+        if (!Imm32CheckAndApplyAppCompat(0, szImeFile))
+            break;
+
+        g_hCtfIme = LoadLibraryW(szImeFile);
+        if (!g_hCtfIme)
+            break;
+
+        bSuccess = TRUE;
+
+#undef DEFINE_CTF_IME_FN
+#define DEFINE_CTF_IME_FN(func_name, ret_type, params) \
+        CTF_IME_FN(func_name) = (FN_##func_name)GetProcAddress(g_hCtfIme, #func_name); \
+        if (!CTF_IME_FN(func_name)) \
+        { \
+            bSuccess = FALSE; \
+            break; \
+        }
+#include "CtfImeTable.h"
+    } while (0);
+
+    if (g_hCtfIme && !bSuccess)
+    {
+#undef DEFINE_CTF_IME_FN
+#define DEFINE_CTF_IME_FN(func_name, ret_type, params) \
+        CTF_IME_FN(func_name) = NULL;
+#include "CtfImeTable.h"
+
+        FreeLibrary(g_hCtfIme);
+        g_hCtfIme = NULL;
+    }
+
+    RtlLeaveCriticalSection(&gcsImeDpi);
+    return g_hCtfIme;
 }
 
-// Win: Internal_CtfImeDestroyInputContext
-HRESULT APIENTRY Imm32CtfImeDestroyInputContext(HIMC hIMC)
+HRESULT Imm32CtfImeCreateThreadMgr(VOID)
 {
-    if (!Imm32LoadCtfIme())
+    if (!Imm32LoadCtfIme() || CTF_IME_FN(CtfImeCreateThreadMgr) == NULL)
         return E_FAIL;
+    return CTF_IME_FN(CtfImeCreateThreadMgr)();
+}
 
-#if 1
-    FIXME("(%p)\n", hIMC);
-    return E_NOTIMPL;
-#else
-    return g_pfnCtfImeDestroyInputContext(hIMC);
-#endif
+HRESULT Imm32CtfImeCreateInputContext(HIMC hIMC)
+{
+    if (!Imm32LoadCtfIme() || CTF_IME_FN(CtfImeCreateInputContext) == NULL)
+        return E_FAIL;
+    return CTF_IME_FN(CtfImeCreateInputContext)(hIMC);
+}
+
+static BOOL CALLBACK Imm32EnumIMC(HIMC hIMC, LPARAM lParam)
+{
+    Imm32CtfImeCreateInputContext(hIMC);
+    return TRUE;
+}
+
+HRESULT Imm32CtfImeDestroyInputContext(HIMC hIMC)
```

Comment:
Since all these are new functions, could you use our formatting with SAL annotations?
Also, doxygen documentation for each would be much welcome, because: 1. All that code is weirdly _not_ commented, and 2. Those functions are undocumented, 3. nobody knows what they are for, except maybe by you.

Reply:
Added SAL annotations.
Added comments.